README
IguanaIR software / firmware project README
===========================================

This project provides the software and firmware used with the
IguanaWorks usbir hardware.  This hardware is used to communicate via
infrared signals between remote controls, computer systems, and
consumer electronics.  The hardware is most commonly used along side
LIRC or WinLirc by individuals building customized media centers or
similar projects, but people have also controlled glowing juggling
balls or used these devices in university projects.

This code currently supports multiple Linux distributions on PCs or
embeded platforms, such as the Raspberry Pi, along with Windows 7
through Windows 10.  Windows XP and Vista should also work, but are
not explicitly supported at this time.  This software includes
firmware update tools, command line testing tools, and Python (version
2 or 3) and C programming APIs for people who want to write their own
code to interact with devices controlled by infrared signals.

Alternatively, a Linux kernel driver also supports the IguanaWorks
usbir hardware and does not require or use the software provided here.

With either this software, or the kernel driver, Linux users commonly
rely on LIRC to decode the infrared signals into a more user-friendly
format such as "TV Power" or "Heater On":

  http://www.lirc.org/

Windows users usually rely on WinLIRC for signal decoding and coupled
with plugins for frameworks such as Girder or EventGhost these
configurations can tie infrared signals into home automation:

  http://winlirc.sourceforge.net/
  https://www.promixis.com/girder.php
  http://www.eventghost.org/



The hardware can be purchased from our website and bulk discounts are
available:

  http://www.iguanaworks.net

The IguanaWorks website also provides Windows binaries, frequently
asked questions, and a support site to contact the developers with
questions, issues, or descriptions of interesting projects.

Thank you for your interest.
 -The IguanaWorks Team


ChangeLog
Version: 1.2.1
Author:  Iguanaworks Support <support@iguanaworks.net>
When:    Sun, 20 Nov 2022 11:12:38 -0600
Debian:  jammy, 1
Fedora:  1
Message:
  * Update Build requirements for newer .deb based distros
END

Version: 1.2.0
Author:  Iguanaworks Support <support@iguanaworks.net>
When:    Sat, 29 Jul 2017 14:22:28 -0600
Debian:  zesty, 1
Fedora:  1
Message:
  * Big update.
  * New Packaging
END

Version: 1.1.0
Author:  Iguanaworks Support <support@iguanaworks.net>
When:    Sun, 10 Mar 2013 13:28:28 -0600
Debian:  stable, 1
Fedora:  1
Message:
  * Switched from autoconf to cmake
  * Added a device reboot in reflasher
  * svn version numbers built into the executables
END

Version: 1.0.5
Author:  Iguanaworks Support <support@iguanaworks.net>
When:    Tue, 26 Jan 2013 11:00:14 -0700
Debian:  stable, 1
Fedora:  1
Message:
  * 1.0.5 Iguanaworks Release
  * Bug fix on 1.0.4 with new socket location /var/run.
END

Version: 1.0.4
Author:  Iguanaworks Support <support@iguanaworks.net>
When:    Sun, 13 Jan 2013 15:00:14 -0700
Debian:  stable, 1
Fedora:  1
Message:
  * 1.0.4 Iguanaworks Release
  * Various packaging improvement to be standards compliant.
END

Version: 1.0.3
Author:  Iguanaworks Support <support@iguanaworks.net>
When:    Tue, 21 Aug 2012 20:01:44 -0600
Debian:  precise, 1
Fedora:  1
Message:
  * 1.0.3 Iguanaworks Release
  * 0308 Firmware and driver support for 0308 firmware
  * Restart on device disconnect fixes issues on some systems with usb power management problems.
END

Version: 1.0.1
Author:  IguanaWorks Support <support@iguanaworks.net>
When:    Sat, 20 Nov 2010 20:25:29 -0700
Debian:  maverick, 1
Fedora:  1
Message:
  * 1.0.1 Iguanaworks Release
END

Version: 1.0pre2
Author:  Iguanaworks Support <support@iguanaworks.net>
When:    Sun, 08 Nov 2009 08:48:53 -0600
Debian:  jaunty, 1
Fedora:  1
Message:
  * Works with Ubuntu 9.10. Supports slot device, bug fixes
END

Version: 1.0pre1
Author:  Iguanaworks Support <support@iguanaworks.net>
When:    Sun, 12 Jul 2009 18:07:25 -0400
Debian:  jaunty, 1
Fedora:  1
Message:
  * Pre Release of 1.0.
END

Version: 0.99.1
Author:  Iguanaworks Support <support@iguanaworks.net>
When:    Sun, 23 Nov 2008 16:36:37 -0500
Debian:  gutsy, 4
Fedora:  4
Message:
  * 0.99 IguanaWorks Release
END

Version: 0.98.1
Author:  IguanaWorks Support <support@iguanaworks.net>
When:    Mon, 18 Aug 2008 20:42:17 -0600
Debian:  unstable, 1
Fedora:  1
Message:
  * Upgrade to the 0.98 source.
  * Fixed some debian specific packaging dependencies
END

Version: 0.97.1
Author:  IguanaWorks Support <support@iguanaworks.net>
When:    Mon, 18 Aug 2008 20:35:47 -0600
Debian:  unstable, 1
Fedora:  1
Message:
  * Upgraded to the 0.97 source (fixing udev)
END

Version: 0.96
Author:  Iguanaworks Support <support@iguanaworks.net>
When:    Fri Jun 27 2008
Debian:  stable, 1
Fedora:  1
Message:
  * Bug fix release.
END

Version: 0.95.1
Author:  IguanaWorks Support <support@iguanaworks.net>
When:    Fri, 27 Mar 2008 23:28:45 +1000
Debian:  unstable, 1
Fedora:  1
Message:
  * Upgraded to the 0.95 source (fixing udev)
END

Version: 0.94.1
Author:  IguanaWorks Support <support@iguanaworks.net>
When:    Mon, 24 Mar 2008 07:23:45 +1000
Debian:  unstable, 1
Fedora:  1
Message:
  * Upgraded to the 0.94 source.
  * Better windows support
  * a pile of bugs fixed
  * Works with newer firmwares (version 0x102) including frequency and channel support with or without LIRC.
END

Version: 0.93.1
Author:  IguanaWorks Support <support@iguanaworks.net>
When:    Mon, 28 Feb 2008 18:01:45 +1000
Debian:  unstable, 5
Fedora:  5
Message:
  * Fixed some packaging problems.
END

Version: 0.93.1
Author:  IguanaWorks Support <support@iguanaworks.net>
When:    Thu, 14 Jun 2007 00:17:37 -0400
Debian:  unstable, 2
Fedora:  2
Message:
  * Fix good packaging.
END

Version: 0.31
Author:  Iguanaworks Support <support@iguanaworks.net>
When:    Sat Mar 10 2007
Debian:  unstable, 1
Fedora:  1
Message:
  * First release with tentative win32 and darwin support.  Darwin needs some work, and windows needs to interface with applications.
END

Version: 0.30
Author:  Iguanaworks Support <support@iguanaworks.net>
When:    Thu Feb 1 2007
Debian:  unstable, 1
Fedora:  1
Message:
  * Added a utility to change the frequency on firmware version 3, and had to make iguanaRemoveData accessible to python code.
END

Version: 0.29
Author:  Iguanaworks Support <support@iguanaworks.net>
When:    Sun Jan 21 2007
Debian:  unstable, 1
Fedora:  1
Message:
  * Last currently known problem in the driver.  Using clock_gettime instead of gettimeofday to avoid clock rollbacks.
END

Version: 0.26
Author:  Iguanaworks Support <support@iguanaworks.net>
When:    Sun Dec 31 2006
Debian:  unstable, 1
Fedora:  1
Message:
  * Happy New Years! and a bugfix.  Long standing bug that caused the igdaemon to hang is fixed.
END

Version: 0.25
Author:  Iguanaworks Support <support@iguanaworks.net>
When:    Sun Dec 10 2006
Debian:  unstable, 1
Fedora:  1
Message:
  * The socket specification accept a path instead of just an index or label.
END

Version: 0.24
Author:  Iguanaworks Support <support@iguanaworks.net>
When:    Wed Dec 6 2006
Debian:  unstable, 1
Fedora:  1
Message:
  * Fixes bad argument parsing in igdaemon, and the init script *should* work for fedora and debian now.
END

Version: 0.19
Author:  Iguanaworks Support <support@iguanaworks.net>
When:    Wed Oct 18 2006
Debian:  unstable, 1
Fedora:  1
Message:
  * A real release has been made, and we will try to keep track of version numbers a bit better now.
END

Version: 0.10
Author:  Iguanaworks Support <support@iguanaworks.net>
When:    Sat Sep 23 2006
Debian:  unstable, 1
Fedora:  1
Message:
  * Preparing for a real release.
END

Version: 0.9
Author:  Iguanaworks Support <support@iguanaworks.net>
When:    Tue Jul 11 2006
Debian:  unstable, 1
Fedora:  1
Message:

lib/usb_descr.asm
;;*****************************************************************************
;;*****************************************************************************
;;  FILENAME: USB_descr.asm
;;   Version: 1.5, Updated on 2005/08/17 at 15:01:28
;;  Generated by PSoC Designer ver 4.2  b1013 : 02 September, 2004
;;
;;  DESCRIPTION: USB User Module Descriptors
;;
;;  NOTE: User Module APIs conform to the fastcall convention for marshalling
;;        arguments and observe the associated "Registers are volatile" policy.
;;        This means it is the caller's responsibility to preserve any values
;;        in the X and A registers that are still needed after the API
;;        function returns. Even though these registers may be preserved now,
;;        there is no guarantee they will be preserved in future releases.
;;-----------------------------------------------------------------------------
;;  Copyright (c) Cypress Semiconductor 2004, 2005. All Rights Reserved.
;;*****************************************************************************
;;*****************************************************************************

include "m8c.inc"
include "USB.inc"
include "USB_macros.inc"

WIZARD: equ 1
WIZARD_DEFINED_REPORT_STORAGE:    EQU    1
   ;--------------------------------------------------- 
   ;@PSoC_UserCode_BODY_1@ (Do not change this line.)
   ;---------------------------------------------------
   ; Insert your custom code below this banner
   ;---------------------------------------------------
   ; Redefine the WIZARD equate to 0 below by 
   ; uncommenting the WIZARD: equ 0 line 
   ; to allow your custom descriptor to take effect
   ;--------------------------------------------------- 

   ; WIZARD: equ 0
   ; WIZARD_DEFINED_REPORT_STORAGE:    EQU    0

   ;---------------------------------------------------
   ; Insert your custom code above this banner
   ;---------------------------------------------------
   ;@PSoC_UserCode_END@ (Do not change this line.)

IF WIZARD
;-----------------------------------------------
; USB Descriptors--Generated by the USB Setup Wizard
;-----------------------------------------------
AREA  UserModules     (ROM,REL,CON)
;----------------------------------------------------------------------
; Descriptor Dispatch Tables
;   The following tables provide the dispatching pointers and
;   descriptor lengths necessary for the request handler to
;   respond to USB requests.
;----------------------------------------------------------------------

;----------------------------------------------------------------------
; Device Dispatch Table
;   for User Module: (USB)
;----------------------------------------------------------------------
AREA  func_lit     (ROM,REL,CON)
EXPORT USB_DEVICE_LOOKUP
.LITERAL
USB_DEVICE_LOOKUP:                               ;
  LT_START    1                                  ; Number of devices defined for USB
  LT_ENTRY    USB_D0_CONFIG_LOOKUP, USB_D0_CONFIG_DESCR_TABLE
.ENDLITERAL

;----------------------------------------------------------------------
; Configuration Dispatch Table
;   for Device: (USB_D0)
;----------------------------------------------------------------------
AREA  func_lit     (ROM,REL,CON)
EXPORT USB_D0_CONFIG_LOOKUP
.LITERAL
USB_D0_CONFIG_LOOKUP:                            ;
  LT_START    1                                  ; Number of configurations
  LT_ENTRY    USB_D0_C1_EP_SETUP, USB_D0_C1_INTERFACE_LOOKUP
.ENDLITERAL

;TODO: why did we have to add this?
;----------------------------------------------------------------------
; Interface Look Up Table
; 
; This look up table points to three tables.  Each of those tables is
; indexed by the selected interface number provided by the request.
;
; Item 0: HID Report GET_REPORT/SET_REPORT Lookup Table
; Item 1: HID Report Descriptor Transfer Descriptor Table
; Item 2: HID Class Descriptor Transfer Descriptor Table
;----------------------------------------------------------------------
EXPORT USB_D0_C1_INTERFACE_LOOKUP
.LITERAL
USB_D0_C1_INTERFACE_LOOKUP:                      ;
  DW    USB_D0_C1_INTERFACE_RPT_LOOKUP           ; Interface GET_REPORT/SET_REPORT Lookup Table
  DW    USB_D0_C1_HID_CLASS_DESCR_TABLE          ; HID Class Descriptor Transfer Descriptor Table
.ENDLITERAL

;----------------------------------------------------------------------
; Endpoint Setup Table
; 
; This table provides the data to configure the endpoint mode registers
; for IN/OUT direction.
;----------------------------------------------------------------------
AREA  func_lit     (ROM,REL,CON)
.LITERAL
USB_D0_C1_EP_SETUP:                              ;
  DB    USB_DIR_IN                               ; Endpoint EP1(IN)
  DB    USB_DIR_OUT                              ; Endpoint EP2(OUT)
.ENDLITERAL

;----------------------------------------------------------------------
; Interface Lookup Table
; 
; This table is indexed by interface number.
;
;----------------------------------------------------------------------
EXPORT USB_D0_C1_INTERFACE_RPT_LOOKUP
.LITERAL
USB_D0_C1_INTERFACE_RPT_LOOKUP:                  ;
  DW    0                                        ; HID Class Reports not defined for this interface
.ENDLITERAL

;----------------------------------------------------------------------
; HID Class Descriptor transfer descriptor table
;----------------------------------------------------------------------
AREA  func_lit     (ROM,REL,CON)
EXPORT USB_D0_C1_HID_CLASS_DESCR_TABLE
.LITERAL
USB_D0_C1_HID_CLASS_DESCR_TABLE:                 ;
  TD_START_TABLE 1                               ; Number of interfaces/HID Class Descriptors
  TD_ENTRY       USB_DS_ROM, DESCR_SIZE_HID_CLASS, USB_D0_C1_I0_HID_DESCR_START, NULL_PTR
.ENDLITERAL

;----------------------------------------------------------------------
; Configuration Descriptor Table
;  for (USB_D0)
;
;	This table provides transfer descriptors for each USB Configuration
; Descriptor
;----------------------------------------------------------------------
AREA  func_lit     (ROM,REL,CON)
.LITERAL
USB_D0_CONFIG_DESCR_TABLE:                       ;
  TD_START_TABLE 1                               ; Number of configurations
  TD_ENTRY       USB_DS_ROM, USB_D0_C1_DESCR_SIZE, USB_D0_C1_DESCR_START, NULL_PTR
.ENDLITERAL

;----------------------------------------------------------------------
; Device Descriptor Table
;
;	This table provides transfer descriptors for each USB Device Descriptor
;----------------------------------------------------------------------
AREA  func_lit     (ROM,REL,CON)
EXPORT USB_DEVICE_DESCR_TABLE
.LITERAL
USB_DEVICE_DESCR_TABLE:                          
  TD_START_TABLE 1                               ; Number of devices
  TD_ENTRY       USB_DS_ROM, USB_D0_DESCR_SIZE, USB_D0_DESCR_START, NULL_PTR
.ENDLITERAL
;----------------------------------------------------------------------
;----------------------------------------------------------------------
; Interface 0 RAM Allocation (Reused across configurations)
;----------------------------------------------------------------------
IF WIZARD_DEFINED_REPORT_STORAGE
AREA  bss     (RAM,REL,CON)
;----------------------------------------------------------------------
; RAM allocation for IN Reports.
;----------------------------------------------------------------------
EXPORT _USB_INTERFACE_0_IN_RPT_DATA
EXPORT USB_INTERFACE_0_IN_RPT_DATA
_USB_INTERFACE_0_IN_RPT_DATA:                    
USB_INTERFACE_0_IN_RPT_DATA:                     
  BLK  8                                         ;
;----------------------------------------------------------------------
; RAM allocation for OUT Reports.
;----------------------------------------------------------------------
EXPORT _USB_INTERFACE_0_OUT_RPT_DATA
EXPORT USB_INTERFACE_0_OUT_RPT_DATA
_USB_INTERFACE_0_OUT_RPT_DATA:                   
USB_INTERFACE_0_OUT_RPT_DATA:                    
  BLK  8                                         ;
;----------------------------------------------------------------------
; RAM allocation for FEATURE Reports.
;----------------------------------------------------------------------
EXPORT _USB_INTERFACE_0_FEATURE_RPT_DATA
EXPORT USB_INTERFACE_0_FEATURE_RPT_DATA
_USB_INTERFACE_0_FEATURE_RPT_DATA:               
USB_INTERFACE_0_FEATURE_RPT_DATA:                
  BLK  8                                         ;
;----------------------------------------------------------------------
ENDIF  ; WIZARD_DEFINED_REPORT_STORAGE
;----------------------------------------------------------------------
; Device Descriptor (USB_D0)
; This marks the beginning of the Device Descriptor.  This descriptor
; concatenates all of the descriptors in the following format:
;	Device Descriptor
;		Configuration Descriptor (1)
;	  	Interface 1 Descriptor
;				HID Descriptor 1
;			  	Endpoint Descriptor(s)
;			Interface 2 Descriptor
;				HID Descriptor 2
;					Endpoint Descriptor(s)
;		Configuration Descriptor (2)
;			Interface/HID/Endpoint...
;----------------------------------------------------------------------
AREA  UserModules     (ROM,REL,CON)
.LITERAL
USB_D0_DESCR_START:                              ;
  DB    18                                       ; Device Descriptor Length (18)
  DB    1                                        ; bDescriptorType: DEVICE
  DWL   0200H                                    ; bcdUSB (ver 2.0)
  DB    0                                        ; bDeviceClass
  DB    0                                        ; bDeviceSubClass 
  DB    0                                        ; bDeviceProtocol 
  DB    8                                        ; bMaxPacketSize0
  DWL   1781H                                    ; idVendor
  DWL   0938H                                    ; idProduct
  DWL   0000H                                    ; bcdDevice
  DB    STR_HASH_2                               ; iManufacturer
  DB    STR_HASH_3                               ; iProduct
  DB    STR_HASH_0                               ; iSerialNumber
  DB    1                                        ; bNumConfigurations
USB_D0_DESCR_END:                                ;
USB_D0_DESCR_SIZE:  EQU    USB_D0_DESCR_END - USB_D0_DESCR_START

;----------------------------------------------------------------------
; Configuration Descriptor (USB_D0_C1)
; This marks the beginning of the Configuration Descriptor.  This descriptor
; concatenates all of the descriptors in the following format:
; Configuration Descriptor
;	Interface 1 Descriptor
;		HID Descriptor 1
;			Endpoint Descriptor(s)
;	Interface 2 Descriptor
;		HID Descriptor 2
;			Endpoint Descriptor(s)
;----------------------------------------------------------------------
USB_D0_C1_DESCR_START:                           ;
  DB    9                                        ; Configuration Descriptor Length (9)
  DB    2                                        ; bDescriptorType: CONFIGURATION
  DWL   USB_D0_C1_DESCR_SIZE                     ; wTotalLength
  DB    1                                        ; bNumInterfaces
  DB    1                                        ; bConfigurationValue
  DB    STR_HASH_0                               ; iConfiguration
  DB    80H | (0 << 6) | (0 << 5)                ; bmAttributes >
  DB    50                                       ; bMaxPower
;----------------------------------------------------------------------
; Interface Descriptor (USB_D0_C1_I0)
;----------------------------------------------------------------------
  DB    9                                        ; Interface Descriptor Length (9)
  DB    4                                        ; bDescriptorType: INTERFACE
  DB    0                                        ; bInterfaceNumber (zero based)
  DB    0                                        ; bAlternateSetting
  DB    2                                        ; bNumEndpoints
  DB    0                                        ; bInterfaceClass
  DB    0                                        ; bInterfaceSubClass
  DB    0                                        ; bInterfaceProtocol
  DB    STR_HASH_0                               ; iInterface
;----------------------------------------------------------------------
; HID Class Descriptor (USB_D0_C1_I0)
;----------------------------------------------------------------------
USB_D0_C1_I0_HID_DESCR_START:                    
  DB    9                                        ; bLength--HID Class Descriptor Length (9)
  DB    0x21                                     ; bDescriptorType: HID Class
  DWL   0x0111                                   ; bcdHID
  DB    0                                        ; bCountryCode
  DB    1                                        ; bNumDescriptors
  DB    34                                       ; bDescriptorType
  DWL   0                                        ; wDescriptorLength (No report was selected in the setup wizard)
;----------------------------------------------------------------------
; Endpoint Descriptor (USB_D0_C1_I0_E0)
;----------------------------------------------------------------------
  DB    7                                        ; Endpoint Descriptor Length (7)
  DB    5                                        ; bDescriptorType: ENDPOINT
  DB    (EP1 | USB_DIR_IN)                       ; bEndpointAddress
  DB    3                                        ; bmAttributes
  DWL   8                                        ; wMaxPacketSize
  DB    10                                       ; bInterval
;----------------------------------------------------------------------
; Endpoint Descriptor (USB_D0_C1_I0_E1)
;----------------------------------------------------------------------
  DB    7                                        ; Endpoint Descriptor Length (7)
  DB    5                                        ; bDescriptorType: ENDPOINT
  DB    (EP2 | USB_DIR_OUT)                      ; bEndpointAddress
  DB    3                                        ; bmAttributes
  DWL   8                                        ; wMaxPacketSize
  DB    10                                       ; bInterval
;----------------------------------------------------------------------
USB_D0_C1_DESCR_END:
USB_D0_C1_DESCR_SIZE:	EQU	(USB_D0_C1_DESCR_END - USB_D0_C1_DESCR_START)
.ENDLITERAL
;----------------------------------------------------------------------
; HID Class Report Descriptors
;
; This section contains the HID Class Report Descriptors generated
; by the USB User Module Descriptor Generator
;
;----------------------------------------------------------------------
;----------------------------------------------------------------------
; USB String Descriptors
;
; This section contains the USB String Descriptors generated
; by the USB User Module Descriptor Generator
;
; Note:  The string labels are internally generated by the
; descriptor generator
;
; Descriptors that reference string descriptors, use a hashed
; symbol that is set in an EQU directive with each string
; descriptor.
;----------------------------------------------------------------------
STR_HASH_0:  EQU    0                            ; String Hash for the null string
AREA  UserModules     (ROM,REL,CON)
EXPORT USB_StringTable
.LITERAL
USB_StringTable:                                 
  TD_START_TABLE 3                               ; Number of USB Strings
  TD_ENTRY       USB_DS_ROM, LANGID_DESCR_SIZE, LANGID_DESCR_START, NULL_PTR
  TD_ENTRY       USB_DS_ROM, STRING_2_DESCR_SIZE, STRING_2_DESCR_START, NULL_PTR
  TD_ENTRY       USB_DS_ROM, STRING_3_DESCR_SIZE, STRING_3_DESCR_START, NULL_PTR
;----------------------------------------------------------------------
; LANGID Descriptor
;----------------------------------------------------------------------
LANGID_DESCR_START:                              
  DB    ((1 * 2) + 2)                            ; bLength (N+2)
  DB    3                                        ; bDescriptorType (STRING)
  DWL   1033                                     ;
LANGID_DESCR_END:                                
LANGID_DESCR_SIZE:  EQU    (LANGID_DESCR_END - LANGID_DESCR_START)
;----------------------------------------------------------------------
; String Desciptor
;----------------------------------------------------------------------
STR_HASH_2:  EQU    1                            ; String Hash
STRING_2_DESCR_START:                            
  DB    STRING_2_DESCR_SIZE                      ; bLength
  DB    3                                        ; bDescriptorType (STRING)
  DSU   "IguanaWorks"                            ;
STRING_2_DESCR_END:                              
STRING_2_DESCR_SIZE:  EQU    (STRING_2_DESCR_END - STRING_2_DESCR_START)
;----------------------------------------------------------------------
; String Desciptor
;----------------------------------------------------------------------
STR_HASH_3:  EQU    2                            ; String Hash
STRING_3_DESCR_START:                            
  DB    STRING_3_DESCR_SIZE                      ; bLength
  DB    3                                        ; bDescriptorType (STRING)
  DSU   "USB IR Transceiver"                        ;
STRING_3_DESCR_END:                              
STRING_3_DESCR_SIZE:  EQU    (STRING_3_DESCR_END - STRING_3_DESCR_START)
.ENDLITERAL

;-----------------------------------------------
; End of Descriptors
;-----------------------------------------------

ENDIF
   ;---------------------------------------------------
   ;@PSoC_UserCode_BODY_2@ (Do not change this line.)
   ;---------------------------------------------------
   ; Redefine your descriptor table below. You might
   ; cut and paste code from the WIZARD descriptor
   ; above and then make your changes.
   ;---------------------------------------------------

   
   ;---------------------------------------------------
   ; Insert your custom code above this banner
   ;---------------------------------------------------
   ;@PSoC_UserCode_END@ (Do not change this line.)

; End of File USB_descr.asm
